src/data/nat/count.lean

```diff
@@ -0,0 +1,542 @@
+  ext i,
+  rw [mem_filter, mem_range, set.mem_to_finset, set.mem_set_of_eq],
+end
+
+lemma count_monotone : monotone (count p) :=
+λ x y h, list.length_le_of_sublist $ (list.range_sublist.mpr h).filter p
+
+@[simp] lemma count_succ (n : ℕ) : count p (n + 1) = count p n + (if p n then 1 else 0) :=
+by split_ifs; simp [count, list.range_succ, h]
+
+lemma count_add (a b : ℕ) : count p (a + b) = count p a + count (λ k, p (a + k)) b :=
+begin
+  rw [count_eq_card_filter_range, count_eq_card_filter_range, count_eq_card_filter_range, range_add,
+    filter_union, card_disjoint_union, image_filter,
+    card_image_of_injective _ (add_right_injective a)],
+  intros x hx,
+  simp_rw [inf_eq_inter, mem_inter, mem_filter, mem_image, mem_range] at hx,
+  obtain ⟨⟨hx, _⟩, ⟨c, _, rfl⟩, _⟩ := hx,
+  exact (self_le_add_right _ _).not_lt hx,
+end
+
+lemma count_add' (a b : ℕ) : count p (a + b) = count (λ k, p (k + b)) a + count p b :=
+by { rw [add_comm, count_add, add_comm], simp_rw add_comm b, congr' 2 }
+
+lemma count_one : count p 1 = if p 0 then 1 else 0 := by simp
+
+lemma count_succ' (n : ℕ) : count p (n + 1) = count (λ k, p (k + 1)) n + if p 0 then 1 else 0 :=
+by rw [count_add', count_one]
+
+variables {p}
+
+lemma count_succ_eq_succ_count_iff {n : ℕ} : count p (n + 1) = count p n + 1 ↔ p n :=
+by by_cases h : p n; simp [h]
+
+lemma count_succ_eq_count_iff {n : ℕ} : count p (n + 1) = count p n ↔ ¬p n :=
+by by_cases h : p n; simp [h]
+
+alias count_succ_eq_succ_count_iff ↔ _ count_succ_eq_succ_count
+alias count_succ_eq_count_iff ↔ _ count_succ_eq_count
+
+lemma count_le_cardinal (n : ℕ) : (count p n : cardinal) ≤ cardinal.mk (set_of p) :=
+begin
+  obtain h | h := lt_or_le (cardinal.mk (set_of p)) cardinal.omega,
+  { haveI := (cardinal.lt_omega_iff_fintype.mp h).some,
+    rw [cardinal.fintype_card, cardinal.nat_cast_le, count_eq_card_fintype],
+    fapply fintype.card_le_of_injective,
+    exact λ ⟨i, _, hi⟩, ⟨i, hi⟩,
+    tidy },
+  { rw le_antisymm ((cardinal.countable_iff _).mp ((set_of p).countable_encodable)) h,
+    exact (cardinal.nat_lt_omega _).le }
+end
+
+variables {p}
+
+lemma lt_of_count_lt_count {a b : ℕ} (h : count p a < count p b) : a < b :=
+(count_monotone p).reflect_lt h
+
+end count
+
+/-- Find the `n`-th natural number satisfying `p` (indexed from `0`, so `nth p 0` is the first
+natural number satisfying `p`), or `0` if there is no such number. -/
+noncomputable def nth : ℕ → ℕ
+| n := Inf { i : ℕ | p i ∧ ∀ k < n, nth k < i }
+
+instance decidable_pred_mem_set_of [h : decidable_pred p] : decidable_pred (∈ set_of p) := h
+
+lemma nth_zero : nth p 0 = Inf { i : ℕ | p i } := by { rw nth, simp }
+
+@[simp]
+lemma nth_zero_of_zero (h : p 0) : nth p 0 = 0 :=
+begin
+  apply nat.eq_zero_of_le_zero,
+  rw nth,
+  apply nat.Inf_le,
+  simp [h],
+end
+
+lemma nth_zero_of_exists [decidable_pred p] (h : ∃ n, p n) : nth p 0 = nat.find h :=
+by { rw [nth_zero], convert nat.Inf_def h }
+
+lemma nth_set_card_aux {n : ℕ} (hp : set.finite p)
```

Comment:
@YaelDillies Could you fix all of these `set.finite p` and `set.infinite p`?

The issue is that `p` is a predicate, but `set.finite` and `set.infinite` expect a `set`.